### PR TITLE
Package ghcide code actions

### PR DESCRIPTION
**Pull request by berberman**
### Motivation

Previously, functions implement ghcide code actions were required to return `[(T.Text, [TextEdit])]`, and were wired together in `suggestAction`. In plugin handler, we used many shake rules and retrieved values used in concrete implementation of code actions, then called `suggestAction` here, passing those values. However, with more and more code actions implemented by contributors, now `suggestAction` already has 11 parameters, which we need name each of two times and pass all of by hand. So I think it's time to define a record type to gather them together. On the other hand, we introduced another potential return type `Rewrite` to represent the result of the code action and then used `rewrite` to convert it to `TextEdit`, packaging it in `suggestAction`. With this abstraction, it's hard to express some cases that the code action returns either `TextEdit` or `Rewrite`, which depends on the runtime (like `suggestImportDisambiguation` and `suggestNewOrExtendImportForClassMethod`). Thus we could use type class to relax the return type of functions.

### Overview

First of all, we use a new record type `CodeActionArgs` to grab them up:
```haskell
data CodeActionArgs = CodeActionArgs
  { caaExportsMap   :: ExportsMap
  , caaIdeOptions   :: IdeOptions
  , caaParsedModule :: Maybe ParsedModule
  , caaContents     :: Maybe Text
  , caaDf           :: Maybe DynFlags
  , caaAnnSource    :: Maybe (Annotated ParsedSource)
  , caaTmr          :: Maybe TcModuleResult
  , caaHar          :: Maybe HieAstResult
  , caaBindings     :: Maybe Bindings
  , caaGblSigs      :: Maybe GlobalBindingTypeSigsResult
  , caaDiagnostics  :: Diagnostic
  }
```

Then we have a type class `ToCodeAction`:

```haskell
class ToCodeAction a where
  toCodeAction :: CodeActionArgs -> a -> [(T.Text, [TextEdit])]
```

where `a` can be converted into `[(T.Text, [TextEdit])]` (representation of code actions we use in `suggestAction`) under the context of `CodeActionArgs`. And we can define some instances of acceptable return types:

```haskell
instance ToCodeAction [(T.Text, [TextEdit])] where
  toCodeAction _ = id
instance ToCodeAction [(T.Text, [Rewrite])] where
  toCodeAction CodeActionArgs{..} = rewrite caaDf caaAnnSource
instance ToCodeAction [(T.Text, [TextEdit |? Rewrite])] where
  toCodeAction CodeActionArgs{..} r = second (concatMap go) <$> r
   where
    go (InL te) = [te]
    go (InR rw)
      | Just df <- caaDf
        , Just ps <- caaAnnSource
        , Right x <- rewriteToEdit df (annsA ps) rw =
        x
      | otherwise = []
```

For each field `fld` of `CodeActionArgs`, we can make an instance `instance ToCodeAction r => ToCodeAction (fld -> r)` (fields of `CodeActionArgs` don't have overlapped types). In such instances, we take `fld` from `CodeActionArgs` then feed it into `(fld -> r)` to get `r`. If `fld` is `Maybe a`, we make `ToCodeAction (Maybe a -> r)` and `ToCodeAction (a -> r)`. Take `caaDf` as an example:

```haskell
instance ToCodeAction r => ToCodeAction (Maybe DynFlags -> r) where
  toCodeAction caa@CodeActionArgs {caaDf = x} f = toCodeAction caa $ f x
instance ToCodeAction r => ToCodeAction (DynFlags -> r) where
  toCodeAction caa@CodeActionArgs {caaDf = Just x} f = toCodeAction caa $ f x
  toCodeAction _ _  = []
```

Once we have those instances, suggestX (functions providing code actions) could be free to use values from `CodeActionArgs` implicitly, with `Maybe` or without `Maybe`. And we keep backward compatibility -- this PR doesn't change any these kind of functions. Moreover, they could use `[TextEdit |? Rewrite]` as the return type, which addresses the above issue.

### CODE REVIEW

**Comment by pepeiborra** [comment]
There are many ways to skin this cat, I just don't know if this is the best one. 

Maybe it would help if you updated the PR description with an overview of the goals and rationale behind this change (and move the other remarks about `newImport` and `Rewrite` to a separate issue).

**Comment by pepeiborra** [comment]
I haven't had time to review this yet, but I think the new description is awesome

File: ghcide/src/Development/IDE/Plugin/CodeAction/Args/TH.hs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+mkInstances :: Name -> DecsQ
+mkInstances tyConName =
+  reify tyConName >>= \case
+    (TyConI (DataD _ _ _ _ [RecC dataConName tys] _)) -> concat <$> mapM (genForVar dataConName) tys
+    _ -> error "unsupported"
+ where
+  clsType = conT $ mkName "ToCodeAction"
+  methodName = mkName "toCodeAction"
+  tempType = varT $ mkName "r"
+  commonFun dataConName fieldName = funD methodName [clause [mkName "caa" `asP` recP dataConName [fieldPat fieldName $ varP (mkName "x")], varP (mkName "f")] (normalB [|$(varE methodName) caa $ f x|]) []]
```
Comment (pepeiborra) [inline ghcide/src/Development/IDE/Plugin/CodeAction/Args/TH.hs:16]: could you format this file to fix these long lines?

File: ghcide/src/Development/IDE/Plugin/CodeAction/Args.hs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+wrap :: ToCodeAction a => a -> SomeAction
+wrap = SomeAction
+
+unwrap :: CodeActionArgs -> SomeAction -> [(T.Text, [TextEdit])]
+unwrap caa (SomeAction x) = toCodeAction caa x
+
+class ToCodeAction a where
+  toCodeAction :: CodeActionArgs -> a -> [(T.Text, [TextEdit])]
+
```
Comment (pepeiborra) [inline ghcide/src/Development/IDE/Plugin/CodeAction/Args.hs:64]: Could you document this class using parts of the Description of this PR?

**Comment by pepeiborra** [comment]
Now that I understand what is going on, I think this is very cool. Nice Job!

File: ghcide/src/Development/IDE/Plugin/CodeAction.hs [hunk 3]
```diff
@@ -129,55 +125,29 @@ mkCA :: T.Text -> [Diagnostic] -> WorkspaceEdit -> (Command |? CodeAction)
...
-    ] ++ concat
-    [  suggestNewDefinition ideOptions pm text diag
-    ++ suggestNewImport packageExports pm diag
-    ++ suggestDeleteUnusedBinding pm text diag
-    ++ suggestExportUnusedTopBinding text pm diag
-    | Just pm <- [parsedModule]
-    ] ++
-    suggestFillHole diag                   -- Lowest priority
+    [ wrap $ suggestSignature True
+    , wrap suggestExtendImport
```
Comment (pepeiborra) [inline ghcide/src/Development/IDE/Plugin/CodeAction.hs:133]: Do we really really really need an existential wrapper? Wouldn't something simpler work as well?

```suggestion
    , toCodeAction caa suggestExtendImport
```

File: ghcide/src/Development/IDE/Plugin/CodeAction/Args.hs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+--
+-- and
+--
+-- @
+-- instance ToCodeAction r => ToCodeAction (DynFlags -> r) where
+--   toCodeAction caa@CodeActionArgs {caaDf = Just x} f = toCodeAction caa $ f x
+--   toCodeAction _ _  = []
+-- @
+-- will be derived from 'caaDf'.
+mkInstances ''CodeActionArgs
```
Comment (pepeiborra) [inline ghcide/src/Development/IDE/Plugin/CodeAction/Args.hs:104]: Do we really need the TH? The set of instances is unlikely to grow much more, couldn't we just splice them out and remove or comment out the TH ?

File: ghcide/src/Development/IDE/Plugin/CodeAction/Args.hs [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+  toCodeAction _ = id
+
+instance ToCodeAction [(T.Text, [Rewrite])] where
+  toCodeAction CodeActionArgs{..} = rewrite caaDf caaAnnSource
+
+instance ToCodeAction r => ToCodeAction (ParsedSource -> r) where
+  toCodeAction caa@CodeActionArgs{caaAnnSource = Just ps} f = toCodeAction caa $ f $ astA ps
+  toCodeAction _ _ = []
+
+instance ToCodeAction [(T.Text, [TextEdit |? Rewrite])] where
```
Comment (pepeiborra) [inline ghcide/src/Development/IDE/Plugin/CodeAction/Args.hs:75]: isn't `|?` just [`Sum`](https://hackage.haskell.org/package/base-4.14.1.0/docs/Data-Functor-Sum.html#t:Sum) which is available in the Prelude?

Comment (Ailrun) [reply ghcide/src/Development/IDE/Plugin/CodeAction/Args.hs:75]: I think it's more like `Either` with special JSON treatment, as specified at https://github.com/wz1000/haskell-lsp/blob/master/lsp-types/src/Language/LSP/Types/Common.hs

Comment (pepeiborra) [reply ghcide/src/Development/IDE/Plugin/CodeAction/Args.hs:75]: Is the special JSON treatment relevant here? If not, why not use `Either`?

Comment (berberman) [reply ghcide/src/Development/IDE/Plugin/CodeAction/Args.hs]: No. I thought commonly `Left` was used to express the error, so I chose the in scope `|?`. But now I think `Either TextEdit Rewrite` makes sense: we can't express this modification in `Rewrite`, thus we have to use `TextEdit` as the fallback.

File: ghcide/src/Development/IDE/Plugin/CodeAction/Args.hs [hunk 6]
```diff
@@ -21,22 +20,46 @@ import           Development.IDE.Plugin.TypeLenses            (GlobalBindingType
...
+    caaParsedModule :: Maybe ParsedModule,
+    caaContents     :: Maybe T.Text,
+    caaDf           :: Maybe DynFlags,
+    caaAnnSource    :: Maybe (Annotated ParsedSource),
+    caaTmr          :: Maybe TcModuleResult,
+    caaHar          :: Maybe HieAstResult,
+    caaBindings     :: Maybe Bindings,
+    caaGblSigs      :: Maybe GlobalBindingTypeSigsResult,
+    caaDiagnostics  :: Diagnostic
   }
```
Comment (berberman) [inline ghcide/src/Development/IDE/Plugin/CodeAction/Args.hs]: I run ormolu first and then used stylish-haskell, but the later one didn't revert the changes done by the former one... Is that what we expect? @Ailrun

Comment (Ailrun) [reply ghcide/src/Development/IDE/Plugin/CodeAction/Args.hs]: Yes in this case, as record formatting is disabled by default. Maybe it's better to enable it though.

**Pull request merged**